README.md

```diff
@@ -18,189 +18,259 @@ _TODO before entering Phase 2._
+thread spawning should occur. This allows other uses of WebAssembly &mdash;
+i.e., outside the browser &mdash; to specify their own mechanism for spawning
+threads.
 
-// How would this work?
-int pthread_cancel(pthread_t thread);
+[threads proposal]: https://github.com/WebAssembly/threads
+[Web Workers]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers
 
-pthread_t pthread_self(void);
-```
-This is currently highlighted in [[Design choice 2]](#tricky-design-choice-2)
-
-All synchronization functions below can be implemented WASI libc with existing constructs available in the language (atomics) and don't require a new WASI function:  
 
-Mutexes:  
-```
-int pthread_mutex_init(pthread_mutex_t *mutex, const void *attr);
+### Goals
+- __`pthreads` support__: the goal of this proposal is to add the missing
+  functions that are required to implement a subset of `pthreads` API. It does
+  not aim to be identical to the `pthreads` API, but one must be able to create
+  threads that operate on a shared Wasm memory while using the WebAssembly
+  atomic instructions to synchronize on memory access.
 
-int pthread_mutex_lock(pthread_mutex_t *mutex);
+- __library reuse__: standardizing this API would allow re-use of existing
+  libraries and remove friction when porting projects from native
+  execution contexts to WebAssembly and WASI environments (outside the
+  browsers).
 
-int pthread_mutex_unlock(pthread_mutex_t *mutex);
+- __future-compatible__: a possible future direction for WebAssembly is towards
+  supporting multiple threads per instance. This is not possible with the
+  current memory model and instruction set. We aim to expose an API that would
+  be compatible with this future direction.
 
-int pthread_mutex_destroy(pthread_mutex_t *mutex);
+- __browser polyfills__: for browsers, we aim to provide a way to polyfill this
+  API using Web Workers providing similar functionality to what exists in
+  browsers today.
 
-```
 
-RW Locks:  
-```
-int pthread_rwlock_init(pthread_rwlock_t *rwlock, const pthread_rwlockattr_t *attr);
 
-int pthread_rwlock_rdlock(pthread_rwlock_t *rwlock);
+### Non-goals
+- __not POSIX compatible__: this API will not be 100% compatible with all
+  functions and options described by POSIX threads standard.
 
-int pthread_rwlock_tryrdlock(pthread_rwlock_t *rwlock);
+- __no WebAssembly changes__: the current proposal is limited to the WASI APIs
+  signatures and behavior and does not propose changes to the Wasm instruction
+  set.
 
-int pthread_rwlock_trywrlock(pthread_rwlock_t *rwlock);
 
-int pthread_rwlock_wrlock(pthread_rwlock_t *rwlock);
 
-int pthread_rwlock_unlock();
+### API walk-through
 
-int pthread_rwlock_destroy(pthread_rwlock_t *rwlock);
-```
+The API requires the addition of a single function. In pseudo-code:
 
-Conditionals:  
+```C
+status thread_spawn(thread_start_func* start_func, thread_start_arg* start_arg);
```

Comment:
We discussed this a bit in the meeting we had and two things popped out as a result of that:
 - Setting the priority (or even the thread name) is more easily done as a separate API call; this would match how pthreads does it ([`pthread_setschedprio`](https://man7.org/linux/man-pages/man3/pthread_setschedprio.3.html)). This means that in the future we could add a new priority-setting function to wasi-threads and it would not impact spawning. The one requirement we identified to do this is to have a TID that the host can map to the real system thread, but otherwise this kind of feature could be incrementally added.
 - If we did run into a future feature that required changes to the `spawn` signature, we can just use a new name: `wasi_thread_spawn_v2` (or something like that). Presumably wasi-libc could support both names, much like it does now for the `start` and `start_c11` entry points.

I think this means we can avoid the extra parameters until they are absolutely needed, then add them.